contracts/crowdsale/ZxcCrowdsale.sol

```diff
@@ -0,0 +1,278 @@
+
+  uint256 public bonusPresale;
+  uint256 public bonusSale;
+
+  /**
+   * End timestamp to end the crowdsale.
+   */
+  uint256 public endTime;
+
+  /**
+   * Minimum required wei deposit.
+   */
+  uint256 public minimumWeiDeposit;
+
+  /**
+   * Crowdsale cap in ZXC.
+   */
+  uint256 public crowdSaleZxcCap;
+
+  /**
+   * Amount of ZXC tokens sold.
+   */
+  uint256 public zxcSold;
+
+  /**
+   * Address where funds are collected.
+   */
+  address public wallet;
+
+  /**
+   * How many token units buyer gets per wei.
+   */
+  uint256 public rate;
+
+  /**
+   * @dev An event which is triggered when tokens are bought.
+   * @param _from The address sending tokens.
+   * @param _to The address recieving tokens.
+   * @param _weiAmount Purchase amount in wei.
+   * @param _tokenAmount The amount of purchased tokens.
+   */
+  event TokenPurchase(
+    address indexed _from,
+    address indexed _to,
+    uint256 _weiAmount,
+    uint256 _tokenAmount
+  );
+
+  /**
+   * @dev Contract constructor.
+   * @param _walletAddress Address of the wallet which collects funds.
+   * @param _tokenAddress Address of the ZXC token contract.
+   * @param _startTimePresale Start time of presale stage.
+   * @param _startTimeSaleWithBonus Start time of public sale stage with bonus.
+   * @param _startTimeSaleNoBonus Start time of public sale stage with no bonus.
+   * @param _endTime Time when sale ends.
+   * @param _rate ZXC/ETH exchange rate.
+   * @param _crowdSaleZxcCap Number of ZXC tokens offered during the sale.
+   * @param _bonusPresale Bonus token percentage for presale.
+   * @param _bonusSale Bonus token percentage for public sale stage with bonus.
+   * @param _minimumWeiDeposit Minimum required deposit in wei.
+   */
+  constructor(address _walletAddress,
+              address _tokenAddress,
+              uint256 _startTimePresale,  // 1529971200: date -d '2018-06-26 00:00:00 UTC' +%s
+              uint256 _startTimeSaleWithBonus, // 1530662400: date -d '2018-07-04 00:00:00 UTC' +%s
+              uint256 _startTimeSaleNoBonus,  //1530748800: date -d '2018-07-05 00:00:00 UTC' +%s
+              uint256 _endTime,  // 1531872000: date -d '2018-07-18 00:00:00 UTC' +%s
+              uint256 _rate,  // 10000: 1 ETH = 10,000 ZXC
+              uint256 _crowdSaleZxcCap, // 250M
+              uint256 _bonusPresale,  // 10 (%)
+              uint256 _bonusSale,  // 5 (%)
+              uint256 _minimumWeiDeposit  // 1 ether;
+             )
+    public
+  {
+    //uint256 _decimalsMul;
+    uint256 _bonusDividend = 100; // 100%
+
+    require(_walletAddress != address(0));
+    require(_tokenAddress != address(0));
+    require(_tokenAddress != _walletAddress);
+
+    token = Zxc(_tokenAddress);
+
+    uint8 _tokenDecimals = token.decimals();
+    require(_tokenDecimals == 18);  // Sanity check.
+    //_decimalsMul = 10 ** uint256(_tokenDecimals);
+    wallet = _walletAddress;
+
+    require(_bonusPresale > 0);
```

Comment:
Maybe add a max check as well? Goes for _bonusPresale, _bonusSale and rate.

Reply:
Like that it can't be more than 100? That makes sense.

Reply:
cf29223